rfd/0010-api.md

```diff
@@ -0,0 +1,238 @@
+
+    // TLS client is the only client
+    return auth.NewTLSClient(clientConfig)
+```
+
+## Details
+
+### Approach
+
+Keep Teleport's API layer unified, avoid having internal API vs external API layer.
+The access to the API surface is limited by role based access control.
+
+Avoid large refactors, instead use Go's type aliases to move things around while keeping things backwards compatible.
+
+```go
+// Preserve full backwards compatibility
+type services.Role = types.Role
+type auth.Service = proto.Service
+```
+
+#### Move protobuf types and services and generate docs
+
+Point protobuf generators to generate types to another folder and use a system of go submodules.
+
+Use [protoc-gen-doc](https://github.com/pseudomuto/protoc-gen-doc) to generate gRPC API markdown documentation during the protoc build process. This will require us to improve our proto files with user-focused comments.
+
+```bash
+$ ls github.com/gravitational/teleport/api
+# root folder with API specs and is a go submodule
+api/
+  # open API spec used for swagger, see for example K8s Swagger spec
+  # https://github.com/kubernetes/kubernetes/tree/master/api/openapi-spec
+  # These docs can be copied to our doc engine before each doc release, optionally with custom templates.
+  docs/
+    types.md
+    wrappers.md
+    client.md
+    # move generated services types here, while keeping the original proto files
+    # in place in lib/services/services.proto
+  types/
+    types.pb.go
+    wrappers.pb.go
+  # generate auth service lib/auth/proto here
+  client/
+    authservice.pb.go
+```
+
+#### Move client implementation to `api/client`
+
+Move relevant lib/client.*Client code to `api/client`.
+Track dependencies, make sure that new `api/client` module has no dependencies
+on the outside modules anywhere in `lib`.
+
+**Track dependencies**
+
+Track dependencies to be limited to:
+  * `gravitational/roundtrip`,  `gravitational/trace`
+  * grpc google package and its dependencies
+  * protobuf generated code
+  * Keep some auth service specific code like `utils.Addr` out of the `api/client`
+
+**Go native user experience**
+
+Achieve the user experience in Client with a simplified client constructor.
+This should also transparently handle dialing via a proxy.
+
+```go
+import (
+   "github.com/gravitational/teleport/api/client"
+)
+
+func main() {
+   // TLS client is the only TLS client supported in teleport
+   client, err := client.New(client.Config{
+     Addrs: []string{"localhost:3025"},
+     // ContextDialer is optional context dialer
+     ContextDialer: net.DialContext,
+     // direct TLS credentials config
+     Credentials: client.TLSCreds(tls.Config)
+   })
+   ...
+   defer client.Close()
+
+   ping, err := client.Ping()
+   ...
+   fmt.Printf("Ping: %v\n", ping.Version)
+}
+```
+
+```bash
+$ go mod init
+$ go run main.go
+```
+
+**Credential providers**
+
+Support multiple credential providers with chaining
```

Comment:
@klizhentas I want to make sure i understand  what you meant by "chaining" here?

Do you want it to look like
```go
client, err := client.New(client.Config{
     Addrs: []string{"localhost:3025"},
     Credentials: client.ProfileCreds(),
})
```
as opposed to
```go
creds, err := client.ProfileCreds()
if err != nil {
    ...
}

client, err := client.New(client.Config{
     Addrs: []string{"localhost:3025"},
     Credentials: creds,
})
```
I can do that by propagating the error into the credential struct if we want to do that.
```
type Credentials struct {
    tls *tls.Config
    err error
}